### PR TITLE
Experiments with Python interface

### PR DESCRIPTION
**Pull request by timmoon10**
Consider this a series of design experiments, all of which are open to debate:
- There is now a `Layer` class that is distinct from `Module`. A `Layer` is now like a single PyTorch operation (takes an input and produces an output) and a `Module` is now like a PyTorch module (can be applied repeatedly to different inputs). I personally think LBANN uses the word "layer" in a somewhat non-standard way and it should be changed to something like "operation", but that's a question for another day.
- You can specify a `Layer`'s input in the constructor. It's now possible to specify a network in one pass:
```python
conv1 = lp.Convolution('conv1', x, other params)
bn1 = lp.BatchNormalization('bn1', conv1, other params)
relu1 = lp.Relu('relu1', bn1', other params)
```
- The data layout is now an optional parameter when constructing a `Layer`. The default is data-parallel.
- I prefer pure functions over functions with side effects (something something functional programming). To export a `Layer` to prototext, you call `Layer.export_proto()` to get a protobuf layer message and then you are responsible for adding it to the protobuf model message. I've also reworked the layer subclass generator to use a pure function.
- I've reworked the `traverse_model` function, which applies a ~DFS on the layer graph and applies a function to each layer. Now, the `traverse_layer_graph` function (models are more than just their layers) returns a list of layers and you can apply `map` afterwards. In the future, I think it would be best if `traverse_layer_graph` did a DFS on the graph (finding parent nodes in addition to child nodes) and then did a topological sort.
- I removed the base `Module` class since it has nothing to implement, but I don't think this is the right thing to do. It would be better for `Module` to be an abstract base class.
- I've moved ResNet-specific modules within resnet50.py rather than in a globally accessible module, which is closer to PyTorch's approach (https://github.com/pytorch/vision/blob/master/torchvision/models/resnet.py).
- The "parent"/"child" layer terminology has been replaced with "input"/"output". I think this is more natural, but perhaps we want to stick with the terminology we use in the prototext.
- I have a fondness for the conditional operator.

### CODE REVIEW

File: scripts/proto/lbann_proto.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 import sys
 import google.protobuf.text_format
+import collections
```
Comment (ndryden) [inline scripts/proto/lbann_proto.py]: Should be `import collections.abc` (see below).

File: scripts/proto/lbann_proto.py [hunk 2]
```diff
@@ -41,85 +42,78 @@ def _add_to_module_namespace(stuff):
...
-        raise NotImplementedError('proto not implemented')
-
-    def add_weights(self, initializer, name_suffix=None):
-        """Create a new weights object and add it to this module."""
-        if name_suffix is None:
-            name_suffix = '_weights{0}'.format(len(self.weights) + 1)
-        w = Weights(self.name + name_suffix, initializer)
+        # Initialize inputs, outputs, and weights
+        for l in (inputs
+                  if isinstance(inputs, collections.Iterable)
```
Comment (ndryden) [inline scripts/proto/lbann_proto.py]: The check should be `isinstance(inputs, collections.abc.Iterable)`. Using the abstract base classes from `collections` is deprecated (and will be removed in Python 3.8).

File: scripts/proto/lbann_proto.py [hunk 3]
```diff
@@ -41,85 +42,78 @@ def _add_to_module_namespace(stuff):
...
+                  else [inputs]):
+            self.add_input(l)
+        for l in (outputs
+                  if isinstance(outputs, collections.Iterable)
```
Comment (ndryden) [inline scripts/proto/lbann_proto.py]: `collections.abc`

File: scripts/proto/lbann_proto.py [hunk 4]
```diff
@@ -41,85 +42,78 @@ def _add_to_module_namespace(stuff):
...
+                  else [outputs]):
+            self.add_output(output)
+        for w in (weights
+                  if isinstance(weights, collections.Iterable)
```
Comment (ndryden) [inline scripts/proto/lbann_proto.py]: `collections.abc`

File: scripts/proto/lbann_proto.py [hunk 5]
```diff
@@ -41,85 +42,78 @@ def _add_to_module_namespace(stuff):
...
-
-    """
-
-    def __init__(self, name, data_layout):
-        self.name = name
-        self.data_layout = data_layout
-        self.parents = []
-        self.children = []
+    def __init__(self, name='', inputs=[], outputs=[],
+                 weights=[], data_layout=''):
```
Comment (ndryden) [inline scripts/proto/lbann_proto.py]: I like making the data layout optional, but while I'm aware LBANN will default it to `data_parallel` if it's not present, I think it's worth being explicit about it when writing to the prototext. So I would just change the default to `data_layout='data_parallel'`.

File: scripts/proto/lbann_proto.py [hunk 6]
```diff
@@ -371,34 +372,37 @@ def proto(self, model):
     _create_proto_callback(field.message_type.name, field.name)
 _add_to_module_namespace(_proto_callbacks)
 
-def traverse_model(first_module, func):
-    """Traverse a model and apply func to each layer.
+def traverse_layer_graph(start_layers):
```
Comment (ndryden) [inline scripts/proto/lbann_proto.py]: Going this route, I think it would be better to make this a generator instead of explicitly returning a list. It's more Pythonic.

File: scripts/proto/lbann_proto.py [hunk 7]
```diff
@@ -371,34 +372,37 @@ def proto(self, model):
...
-                and all([parent in visited for parent in child.parents])):
-                stack.append(child)
+        for output in l.outputs:
+            if ((output not in visited)
+                and all([(i in visited) for i in output.inputs])):
+                stack.append(output)
+    return layers
 
-def save_model(filename, first_module, data_layout, mini_batch_size, epochs,
+def save_model(filename, layers, mini_batch_size, epochs,
```
Comment (ndryden) [inline scripts/proto/lbann_proto.py]: 1. Let's keep the data layout argument, just make it default to `data_parallel` (as above).
2. I think `first_module` (or `first_layer`, or maybe `first_modules`/`layers` since we could have multiple input layers, I guess) is a better argument name than `layers`, which reads like you need to provide _all_ the layers in the network.

File: scripts/proto/lbann_proto.py [hunk 8]
```diff
@@ -410,7 +414,8 @@ def save_model(filename, first_module, data_layout, mini_batch_size, epochs,
         metric.proto(pb.model)
     for callback in callbacks:
         callback.proto(pb.model)
-    traverse_model(first_module, lambda l: l.proto(pb.model))
+    layers = traverse_layer_graph(layers)
+    pb.model.layer.extend([l.export_proto() for l in layers])
```
Comment (ndryden) [inline scripts/proto/lbann_proto.py:527]: If `traverse_layer_graph` becomes a generator, this can turn into `pb.model.layer.extend([l.export_proto() for l in traverse_layer_graph(layers)])`.

File: scripts/proto/lbann_proto.py [hunk 9]
```diff
@@ -41,85 +42,78 @@ def _add_to_module_namespace(stuff):
...
+    def __init__(self, name='', inputs=[], outputs=[],
+                 weights=[], data_layout=''):
+        Layer.num_layers += 1
+        self.name = name if name else 'layer{0}'.format(Layer.num_layers-1)
+        self.inputs = []
+        self.outputs = []
         self.weights = []
+        self.data_layout = data_layout
 
-    def __call__(self, parent):
```
Comment (ndryden) [inline scripts/proto/lbann_proto.py:63]: I think we should keep the callable syntax for chaining layers together.

File: scripts/proto/lbann_proto.py [hunk 10]
```diff
@@ -41,85 +42,78 @@ def _add_to_module_namespace(stuff):
...
         self.weights.append(w)
-        return w
-
-    def get_weights(self):
-        """Return the weights associated with this module."""
-        return self.weights
 
-    def _add_layer(self, model):
-        """Add a layer entry to model (for use by proto).
+def _create_layer_subclass(type_name, layer_field_name):
```
Comment (ndryden) [inline scripts/proto/lbann_proto.py]: This definitely needed improved documentation, otherwise I'd look at it in a month and it would be black magic. Thanks! 😄

**Review by ndryden — CHANGES_REQUESTED** [review]
Overall I think this is moving in a good direction.

On input/output versus parent/child: I'm not so concerned about the terminology in the prototext as long as we can keep a clear interface. I tend to think this terminology differences depends on whether you think of this as constructing a graph, or feeding layers into each other. Whichever you think is most clear is fine with me for this.

Broader changes:
* I think it's worth keeping the ResNet building blocks in `blocks.py`. ResNets (and their components) are widely-used enough that I think they can go there. In the future, as we collect more building blocks, we can look at splitting them out.
* Let's reintroduce `Module` as an [abstract base class](https://docs.python.org/3/library/abc.html) (it might be worth having `Layer` inherit from it, since a layer is just a module with a single op). It probably needs abstract `export_proto` and `__call__` methods.
* Let's standardize toward the approach you've used for `Layer`, to keep the rest of the API uniform.

We also need to figure out how to handle weights associated with layers. I have some thoughts I'll send you on Slack.

File: scripts/proto/lbann_proto.py (hunk 7 above)
Comment (timmoon10) [reply scripts/proto/lbann_proto.py]: 1. As far as I'm aware of, the `data_layout` field in the `Model` protobuf message doesn't do anything.
2. I think it might be simpler if we expected the user to provide all the layers. It's fairly straightforward with the `traverse_layer_graph` function.

File: scripts/proto/lbann_proto.py (hunk 5 above)
Comment (timmoon10) [reply scripts/proto/lbann_proto.py]: Done.

File: scripts/proto/lbann_proto.py (hunk 2 above)
Comment (timmoon10) [reply scripts/proto/lbann_proto.py]: Done.

File: scripts/proto/lbann_proto.py (hunk 6 above)
Comment (timmoon10) [reply scripts/proto/lbann_proto.py]: Done. Our traversals may be a bit heavy-weight (DFS to find root nodes, DFS for topologically sorted traversal), but I think safety is more important than performance here.

File: scripts/proto/lbann_proto.py (hunk 9 above)
Comment (timmoon10) [reply scripts/proto/lbann_proto.py:63]: Done.

**Comment by timmoon10** [comment]
Ready for another round of review:
- Style changes to `Layer` have been propagated to `Initializer`, `Weights`, `ObjectiveFunctionTerm`, `ObjectiveFunction`, `Metric`, and `Callback`.
- Returned back to "parent"/"child" layer terminology.
- Layer and weight names are now advanced options. Specifying a network looks something like:
```python
conv1 = lp.Convolution(x, other params)
bn1 = lp.BatchNormalization(conv1, other params)
relu1 = lp.Relu(bn1, other params)
```
- `traverse_layer_graph` now works with any start node, not just root nodes. It is a generator function that outputs layers in a topological order. It will probably fail horribly if there is a cycle.
- Restored `Module` and added documentation. It should be used in a similar manner as PyTorch, i.e. a `Module` is a pattern that can be applied repeatedly to different inputs.
- I've documented how our usage of the word "layer" is inconsistent with PyTorch, TensorFlow, and Keras. Their "layers" are similar to our "modules". 
- Implemented fully-connected and convolution modules, which are similar to fully-connected and convolution "layers" in PyTorch/TensorFlow/Keras.

This is ready to merge, ~~although we need to update the readme~~.

File: scripts/proto/README.md [hunk 11]
```diff
@@ -1,166 +1,186 @@
...
 
+_A possible point of confusion_: LBANN uses the term "layer" in a
+somewhat unpopular manner. In LBANN and Caffe, a layer is a single
+operation that is applied to a fixed set of input tensors to obtain a
+fixed set of output tensors. In PyTorch, TensorFlow, and Keras, a
+layer is a pattern of operations that can be applied multiple times to
+different sets of input tensors, each time obtaining a different set
+of output tensors. Their usage of the word "layer" is closer to our
+usage of the word "module." In fact, PyTorch's "layers" are subclasses
+of their `Module` class.
```
Comment (ndryden) [inline scripts/proto/README.md:46]: I would clarify that part of this is because, in this Python interface, we're not really doing forward/back prop, but just setting up the structure of the network.

File: scripts/proto/README.md [hunk 12]
```diff
@@ -1,166 +1,186 @@
...
-Right now, the best source for layers (or similar) and their associated
-arguments is to check the `src/proto/lbann.proto` file. All fields present in
-a message there are supported as keyword arguments in this API.
+Right now, the best source for documentation on neural network
+components (layers, weights, etc.) is in `src/proto/lbann.proto`. All
+fields present in a message are supported as keyword arguments in this
+API.
 
-## Building blocks
+## `lbann_module`
```
Comment (ndryden) [inline scripts/proto/README.md]: Should be `lbann_modules`.

File: scripts/proto/lbann_modules.py [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+        if len(self.weights) == 0:
+            self.weights.append(
+                lp.Weights(initializer=lp.HeNormalInitializer()))
+        if len(self.weights) == 1:
+            self.weights.append(
+                lp.Weights(initializer=lp.ConstantInitializer(value=0.0)))
+
+        # Initialize activation layer
+        self.activation = None
+        if activation:
```
Comment (ndryden) [inline scripts/proto/lbann_modules.py:107]: This prevents using modules as an action (if, for whatever reason, that makes sense). Not sure if we care.

File: scripts/proto/lbann_modules.py [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+                             'but got {0}'.format(len(self.weights)))
+        if len(self.weights) == 0:
+            self.weights.append(
+                lp.Weights(initializer=lp.HeNormalInitializer()))
+        if len(self.weights) == 1:
+            self.weights.append(
+                lp.Weights(initializer=lp.ConstantInitializer(value=0.0)))
+
+        # Initialize activation layer
+        self.activation = None
```
Comment (ndryden) [inline scripts/proto/lbann_modules.py:194]: Ditto on modules.

File: scripts/proto/lbann_proto.py [hunk 15]
```diff
@@ -349,68 +475,73 @@ def __init__(self, **kwargs):
...
-    """Traverse a model and apply func to each layer.
+# ==============================================
+# Export models
+# ==============================================
 
-    first_module is the first module in the model (i.e. the input layer).
-    The model is traversed in a depth-first order except that no layer is
-    visited until all its ancestors have been visited.
+def save_model(filename, mini_batch_size, epochs,
+               layers=[], weights=[], objective_function=None,
```
Comment (ndryden) [inline scripts/proto/lbann_proto.py]: Does it ever make sense to pass no layers? If not, maybe we shouldn't default this to `[]`.

File: scripts/proto/lbann_proto.py [hunk 16]
```diff
@@ -349,68 +475,73 @@ def __init__(self, **kwargs):
...
-    # Add the objective function, metrics, and callbacks.
-    objective_func.proto(pb.model)
-    for metric in metrics:
-        metric.proto(pb.model)
-    for callback in callbacks:
-        callback.proto(pb.model)
-    traverse_model(first_module, lambda l: l.proto(pb.model))
+
+    # Add layers
+    layers = list(traverse_layer_graph(layers))
```
Comment (ndryden) [inline scripts/proto/lbann_proto.py:526]: If we're calling `traverse_layer_graph` here, do we need to pass all the layers in for `layers`? Should document appropriately.

File: scripts/proto/resnet50.py [hunk 17]
```diff
@@ -1,69 +1,139 @@
...
+                     pool_dims_i=7, pool_pads_i=0, pool_strides_i=1,
+                     pool_mode='average')
+fc = lp.FullyConnected(avgpool,
+                       weights=lp.Weights(initializer=lp.HeNormalInitializer()),
+                       hint_layer=labels, has_bias=False)
+softmax = lp.Softmax(fc)
+ce = lp.CrossEntropy([softmax, labels])
+top1 = lp.CategoricalAccuracy([softmax, labels])
+top5 = lp.TopKCategoricalAccuracy([softmax, labels], k=5)
+layers = list(lp.traverse_layer_graph(input))
```
Comment (ndryden) [inline scripts/proto/resnet50.py:118]: `lp.save_model` calls `traverse_layer_graph` internally-- we should do so in one place, not both.

**Review by ndryden — COMMENTED** [review]
I think this looks good, just minor changes. I will test and approve later tonight.

Vaguely related comment: If this becomes the primary approach for creating prototexts, and most people don't explicitly name their layers, we may want to tone down the amount of output related to layers/weights that LBANN produces by default. It's already very extensive for reasonably-sized models (e.g. ResNet-50), and with automatically-generated names is probably incomprehensible.

File: scripts/proto/lbann_proto.py (hunk 15 above)
Comment (timmoon10) [reply scripts/proto/lbann_proto.py]: I could imagine someone doing something with just weights and objective functions, but that would be a very weird use-case.

File: scripts/proto/lbann_proto.py (hunk 16 above)
Comment (timmoon10) [reply scripts/proto/lbann_proto.py:526]: This is just for safety. It's an extra graph traversal, but I assume that that is relatively cheap.

**Comment by timmoon10** [comment]
I agree it would be worthwhile to make the layer/weights printing an advanced feature, either a command-line flag or a callback.

File: scripts/proto/lbann_modules.py (hunk 13 above)
Comment (timmoon10) [reply scripts/proto/lbann_modules.py:107]: Thinking about it, a module is actually more natural here. We are basically mimicking a module that applies an activation function. Perhaps we could implement a module that wraps around a single layer? If the user provides a layer, we can construct the corresponding module. Or maybe it would be messier than it's worth.

File: scripts/proto/resnet50.py (hunk 17 above)
Comment (timmoon10) [reply scripts/proto/resnet50.py:118]: We need the list of layers here so that we can extract their weights and feed them into L2 regularization. I'd like to call `traverse_layer_graph` inside of `save_model` for safety. Maybe I'm being too cautious, but these graphs should not be egregiously large. If they were really so large that graph traversal were painful, imagine how long it would take for a single forward or backward prop iteration.

Comment (ndryden) [reply scripts/proto/resnet50.py:118]: Fair enough. 👍

File: scripts/proto/lbann_modules.py (hunk 13 above)
Comment (ndryden) [reply scripts/proto/lbann_modules.py:107]: That might be more complex than it's worth. Alternately, we could make layers a subclass of modules. This currently seems like an edge-case we can address later though.

**Review by ndryden — APPROVED** [review]

**Comment by ndryden** [comment]
Tested on Lassen, everything seems good for the generated ResNet-50 model.
